src/tree/updater_gpu_hist.cu

```diff
@@ -520,74 +335,87 @@ struct GPUHistMakerDevice {
-          dh::ToSpan(monotone_constraints));
-
-      // Reduce over features to find best feature
-      size_t cub_bytes = 0;
-      cub::DeviceReduce::Reduce(nullptr,
-                                cub_bytes, d_split_candidates.data(),
-                                d_result.data(), d_split_candidates.size(), op,
-                                DeviceSplitCandidate(), streams[i]);
-      dh::TemporaryArray<char> cub_temp(cub_bytes);
-      cub::DeviceReduce::Reduce(reinterpret_cast<void*>(cub_temp.data().get()),
-                                cub_bytes, d_split_candidates.data(),
-                                d_result.data(), d_split_candidates.size(), op,
-                                DeviceSplitCandidate(), streams[i]);
-    }
+    auto sampled_features = column_sampler.GetFeatureSet(0);
+    sampled_features->SetDevice(device_id);
+    common::Span<bst_feature_t> feature_set =
+        interaction_constraints.Query(sampled_features->DeviceSpan(), nidx);
+    auto matrix = page->GetDeviceAccessor(device_id);
+    EvaluateSplitInputs<GradientSumT> inputs{
+        nidx,
+        {root_sum.GetGrad(), root_sum.GetHess()},
+        gpu_param,
+        feature_set,
+        matrix.feature_segments,
+        matrix.gidx_fvalue_map,
+        matrix.min_fvalue,
+        hist.GetNodeHistogram(nidx),
+        node_value_constraints[nidx],
+        dh::ToSpan(monotone_constraints)};
+    EvaluateSingleSplit(dh::ToSpan(splits_out), inputs);
+    std::vector<DeviceSplitCandidate> result(1);
+    dh::safe_cuda(cudaMemcpy(result.data(), splits_out.data().get(),
+                             sizeof(DeviceSplitCandidate) * splits_out.size(),
+                             cudaMemcpyDeviceToHost));
+    return result.front();
+  }
 
-    dh::safe_cuda(cudaMemcpy(result_all.data(), d_result_all.data().get(),
-                             sizeof(DeviceSplitCandidate) * d_result_all.size(),
+  std::vector<DeviceSplitCandidate> EvaluateLeftRightSplits(
+      ExpandEntry candidate, int left_nidx, int right_nidx,
+      const RegTree& tree) {
+    dh::TemporaryArray<DeviceSplitCandidate> splits_out(2);
+    GPUTrainingParam gpu_param(param);
+    auto left_sampled_features =
+        column_sampler.GetFeatureSet(tree.GetDepth(left_nidx));
+    left_sampled_features->SetDevice(device_id);
+    common::Span<bst_feature_t> left_feature_set =
+        interaction_constraints.Query(left_sampled_features->DeviceSpan(),
+                                      left_nidx);
+    auto right_sampled_features =
+        column_sampler.GetFeatureSet(tree.GetDepth(right_nidx));
+    right_sampled_features->SetDevice(device_id);
+    common::Span<bst_feature_t> right_feature_set =
+        interaction_constraints.Query(right_sampled_features->DeviceSpan(),
+                                      left_nidx);
+    auto matrix = page->GetDeviceAccessor(device_id);
+
+    EvaluateSplitInputs<GradientSumT> left{left_nidx,
```

Comment:
Nitpick, can we keep the formatting of `left` and `right` consistent?

Reply:
This is just what clang format decided, is it really a problem? It's quite annoying for me to have manually format one section of code when the rest is automatically formatted.